lib/puppet/provider/sysctl/augeas.rb

```diff
@@ -39,52 +39,72 @@ def self.sysctl_get(key)
 
   confine :feature => :augeas
 
-  def self.instances(reference_resource = nil)
+  def self.instances(reference_resources = nil)
     return @resource_cache if @resource_cache
 
     resources = nil
 
-    augopen(reference_resource) do |aug|
-      resources ||= []
-
-      aug.match("$target/*").each do |spath|
-        resource = {
-          :ensure  => :present,
-          :persist => :true
-        }
-
-        basename = spath.split("/")[-1]
-        resource[:name] = basename.split("[")[0]
-        next unless resource[:name]
-        next if (resource[:name] == "#comment")
-
-        resource[:value] = aug.get("#{spath}")
-
-        # Only match comments immediately before the entry and prefixed with
-        # the sysctl name
-        cmtnode = aug.match("$target/#comment[following-sibling::*[1][self::#{basename}]]")
-        unless cmtnode.empty?
-          comment = aug.get(cmtnode[0])
-          if comment.match(/#{resource[:name]}:/)
-            resource[:comment] = comment.sub(/^#{resource[:name]}:\s*/, "")
+    if reference_resources
+      augopen(reference_resources.first.last) do |aug|
+        resources ||= []
+
+        aug.match("$target/*").each do |spath|
+          resource = {
+            :ensure  => :present,
+            :persist => :true
+          }
+
+          basename = spath.split("/")[-1]
+          resource[:name] = basename.split("[")[0]
+          next unless resource[:name]
+          next if (resource[:name] == "#comment")
+
+          resource[:value] = aug.get("#{spath}")
+
+          # Only match comments immediately before the entry and prefixed with
+          # the sysctl name
+          cmtnode = aug.match("$target/#comment[following-sibling::*[1][self::#{basename}]]")
+          unless cmtnode.empty?
+            comment = aug.get(cmtnode[0])
+            if comment.match(/#{resource[:name]}:/)
+              resource[:comment] = comment.sub(/^#{resource[:name]}:\s*/, "")
+            end
           end
-        end
 
-        resources << resource
+          resources << resource
+        end
       end
     end
 
     # Grab everything else
     resources ||= []
 
-    sysctl_all_args = '-a'
+    sysctl_output = ''
     sep = '='
+    if reference_resources
+      sysctl_args = ['-e']
+
+      if Facter.value(:kernel) == 'FreeBSD'
+        sysctl_args = '-ieW'
+      end
+
+      # Split this into chunks so that we don't exceed command line limits
+      reference_resources.each_slice(30) do |reference_resource_slice|
+        sysctl_args << reference_resource_slice.map { |_ref_name, ref_obj| ref_obj.title }
+
+        sysctl_output += sysctl(sysctl_args.flatten)
+      end
+    else
```

Comment:
Or maybe a short circuit at the beginning of `self.instances` with `unless reference_resources` and these 5 lines, so we can get that out of the way for the rest of the method.

Reply:
I had to split between the file pull (above) and the rest (here). Also, you have to pass `-a` if you want `puppet resource sysctl` to work.

I can probably prune the results from the file pulls as well though.